
<!DOCTYPE html>


<html lang="en" data-content_root="../../../" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Basic analysis methods within the spatialdata ecosystem &#8212; My Jupyter Book</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../../../_static/styles/theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../../../_static/styles/bootstrap.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../../../_static/styles/pydata-sphinx-theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />

  
  <link href="../../../_static/vendor/fontawesome/6.5.2/css/all.min.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../../../_static/vendor/fontawesome/6.5.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../../_static/vendor/fontawesome/6.5.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../../_static/vendor/fontawesome/6.5.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css?v=03e43079" />
    <link rel="stylesheet" type="text/css" href="../../../_static/styles/sphinx-book-theme.css?v=eba8b062" />
    <link rel="stylesheet" type="text/css" href="../../../_static/togglebutton.css?v=13237357" />
    <link rel="stylesheet" type="text/css" href="../../../_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="../../../_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/sphinx-thebe.css?v=4fa983c6" />
    <link rel="stylesheet" type="text/css" href="../../../_static/sphinx-design.min.css?v=95c83b7e" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../../../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b" />
<link rel="preload" as="script" href="../../../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b" />
  <script src="../../../_static/vendor/fontawesome/6.5.2/js/all.min.js?digest=dfe6caa3a7d634c4db9b"></script>

    <script src="../../../_static/documentation_options.js?v=9eb32ce0"></script>
    <script src="../../../_static/doctools.js?v=9a2dae69"></script>
    <script src="../../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../../../_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="../../../_static/copybutton.js?v=f281be69"></script>
    <script src="../../../_static/scripts/sphinx-book-theme.js?v=887ef09a"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="../../../_static/togglebutton.js?v=4a39c7ea"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="../../../_static/design-tabs.js?v=f930bc37"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script async="async" src="../../../_static/sphinx-thebe.js?v=c100c467"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'day_2/spatialdata/beginner/nb4_simple_analysis_using_scverse_tools';</script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-primary-sidebar-checkbox"/>
  <label class="overlay overlay-primary" for="pst-primary-sidebar-checkbox"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-secondary-sidebar-checkbox"/>
  <label class="overlay overlay-secondary" for="pst-secondary-sidebar-checkbox"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../../../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search this book..."
         aria-label="Search this book..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
        
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  
    
  

<a class="navbar-brand logo" href="../../../index.html">
  
  
  
  
  
  
    <p class="title logo__title">My Jupyter Book</p>
  
</a></div>
        <div class="sidebar-primary-item">

 <script>
 document.write(`
   <button class="btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script></div>
        <div class="sidebar-primary-item"><nav class="bd-links bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="../../../index.html">
                    GSCN Workshop 2025
                </a>
            </li>
        </ul>
        
    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><button class="sidebar-toggle primary-toggle btn btn-sm" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</button></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="../../../_sources/day_2/spatialdata/beginner/nb4_simple_analysis_using_scverse_tools.ipynb" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.ipynb</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<script>
document.write(`
  <button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="theme-switch fa-solid fa-sun fa-lg" data-mode="light"></i>
    <i class="theme-switch fa-solid fa-moon fa-lg" data-mode="dark"></i>
    <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"></i>
  </button>
`);
</script>


<script>
document.write(`
  <button class="btn btn-sm pst-navbar-icon search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>
<button class="sidebar-toggle secondary-toggle btn btn-sm" title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</button>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Basic analysis methods within the spatialdata ecosystem</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#subset-the-xenium-data-for-the-scope-of-this-workshop">Subset the Xenium data for the scope of this workshop</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#investigate-the-proposed-leiden-clusters">Investigate the proposed Leiden clusters</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#rename-clusters">Rename clusters</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#studying-the-cell-type-fingerprint-for-regions-of-interest">Studying the cell-type fingerprint for regions of interest</a></li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section class="tex2jax_ignore mathjax_ignore" id="basic-analysis-methods-within-the-spatialdata-ecosystem">
<h1>Basic analysis methods within the <code class="docutils literal notranslate"><span class="pre">spatialdata</span></code> ecosystem<a class="headerlink" href="#basic-analysis-methods-within-the-spatialdata-ecosystem" title="Link to this heading">#</a></h1>
<p>In this notebook, we will show how <code class="docutils literal notranslate"><span class="pre">scanpy</span></code> can be used to perform dimensionality reduction and clustering on Xenium data.</p>
<p>A <code class="docutils literal notranslate"><span class="pre">SpatialData</span></code> object is interoperable with external methods thanks to its modular design that builds upon established data structures.</p>
<p>For instance, in this notebook the decoupling of the geometries and the gene expression annotations for cells allows us to operate on the <code class="docutils literal notranslate"><span class="pre">AnnData</span></code> table using <code class="docutils literal notranslate"><span class="pre">scanpy</span></code>, store back the results in the <code class="docutils literal notranslate"><span class="pre">SpatialData</span></code> object and plot the processed data using <code class="docutils literal notranslate"><span class="pre">spatialdata-plot</span></code>.</p>
<p>⚠️ Adjust the variable below to the data path on your specific workstation.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">data_path</span> <span class="o">=</span> <span class="s2">&quot;../data/&quot;</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">spatialdata</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sd</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">spatialdata_plot</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sdp</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">matplotlib.pyplot</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">plt</span>  <span class="c1"># for multi-panel plots later</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">scanpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sc</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">squidpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sq</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>

<span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="p">[</span><span class="n">sd</span><span class="p">,</span> <span class="n">sdp</span><span class="p">,</span> <span class="n">sc</span><span class="p">,</span> <span class="n">sq</span><span class="p">]:</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">p</span><span class="o">.</span><span class="vm">__name__</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">p</span><span class="o">.</span><span class="n">__version__</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>


<span class="n">sdata_xenium</span> <span class="o">=</span> <span class="n">sd</span><span class="o">.</span><span class="n">read_zarr</span><span class="p">(</span><span class="n">data_path</span> <span class="o">+</span> <span class="s2">&quot;xenium.zarr&quot;</span><span class="p">)</span>
<span class="n">sdata_xenium</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output traceback highlight-ipythontb notranslate"><div class="highlight"><pre><span></span><span class="gt">---------------------------------------------------------------------------</span>
<span class="ne">ModuleNotFoundError</span><span class="g g-Whitespace">                       </span>Traceback (most recent call last)
<span class="n">Cell</span> <span class="n">In</span><span class="p">[</span><span class="mi">2</span><span class="p">],</span> <span class="n">line</span> <span class="mi">1</span>
<span class="ne">----&gt; </span><span class="mi">1</span> <span class="kn">import</span><span class="w"> </span><span class="nn">spatialdata</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sd</span>
<span class="g g-Whitespace">      </span><span class="mi">2</span> <span class="kn">import</span><span class="w"> </span><span class="nn">spatialdata_plot</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sdp</span>
<span class="g g-Whitespace">      </span><span class="mi">3</span> <span class="kn">import</span><span class="w"> </span><span class="nn">matplotlib.pyplot</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">plt</span>  <span class="c1"># for multi-panel plots later</span>

<span class="ne">ModuleNotFoundError</span>: No module named &#39;spatialdata&#39;
</pre></div>
</div>
</div>
</div>
<section id="subset-the-xenium-data-for-the-scope-of-this-workshop">
<h2>Subset the Xenium data for the scope of this workshop<a class="headerlink" href="#subset-the-xenium-data-for-the-scope-of-this-workshop" title="Link to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span>
    <span class="sa">f</span><span class="s2">&quot;Data has </span><span class="si">{</span><span class="n">sdata_xenium</span><span class="o">.</span><span class="n">points</span><span class="p">[</span><span class="s1">&#39;transcripts&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">compute</span><span class="p">()</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s2"> transcript locations.&quot;</span>
<span class="p">)</span>
<span class="n">sdata_xenium</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">render_images</span><span class="p">(</span><span class="s2">&quot;he_image&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Clipping input data to the valid range for imshow with RGB data ([0..1] for floats or [0..255] for integers). Got range [-0.18137255..1.0].
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Data has 12165021 transcript locations.
</pre></div>
</div>
<img alt="../../../_images/d2a8dbe10b2959c7e6bf7873bbe24c7487aed15276f9603ea367fc6a969dcac2.png" src="../../../_images/d2a8dbe10b2959c7e6bf7873bbe24c7487aed15276f9603ea367fc6a969dcac2.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sdata_crop</span> <span class="o">=</span> <span class="n">sdata_xenium</span><span class="o">.</span><span class="n">query</span><span class="o">.</span><span class="n">bounding_box</span><span class="p">(</span>
    <span class="n">axes</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;x&quot;</span><span class="p">,</span> <span class="s2">&quot;y&quot;</span><span class="p">],</span>
    <span class="n">min_coordinate</span><span class="o">=</span><span class="p">[</span><span class="mi">25000</span><span class="p">,</span> <span class="mi">5000</span><span class="p">],</span>
    <span class="n">max_coordinate</span><span class="o">=</span><span class="p">[</span><span class="mi">35000</span><span class="p">,</span> <span class="mi">15000</span><span class="p">],</span>
    <span class="n">target_coordinate_system</span><span class="o">=</span><span class="s2">&quot;global&quot;</span><span class="p">,</span>
<span class="p">)</span>
<span class="nb">print</span><span class="p">(</span>
    <span class="sa">f</span><span class="s2">&quot;Crop has </span><span class="si">{</span><span class="n">sdata_crop</span><span class="o">.</span><span class="n">points</span><span class="p">[</span><span class="s1">&#39;transcripts&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">compute</span><span class="p">()</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s2"> transcript locations.&quot;</span>
<span class="p">)</span>
<span class="n">sdata_crop</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">render_images</span><span class="p">(</span><span class="s2">&quot;he_image&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/Users/tim.treis/anaconda3/envs/spatialdata-workshop/lib/python3.11/functools.py:909: UserWarning: The object has `points` element. Depending on the number of points, querying MAY suffer from performance issues. Please consider filtering the object before calling this function by calling the `subset()` method of `SpatialData`.
  return dispatch(args[0].__class__)(*args, **kw)
Clipping input data to the valid range for imshow with RGB data ([0..1] for floats or [0..255] for integers). Got range [-0.094420604..1.0].
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Crop has 2683201 transcript locations.
</pre></div>
</div>
<img alt="../../../_images/4b86bfd60d62ac18db3ee7f8933ac471f877e5bc51e27632a49fd89cf8811064.png" src="../../../_images/4b86bfd60d62ac18db3ee7f8933ac471f877e5bc51e27632a49fd89cf8811064.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sdata_crop</span><span class="o">.</span><span class="n">tables</span><span class="p">[</span><span class="s2">&quot;table&quot;</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 33727 × 377
    obs: &#39;cell_id&#39;, &#39;transcript_counts&#39;, &#39;control_probe_counts&#39;, &#39;control_codeword_counts&#39;, &#39;unassigned_codeword_counts&#39;, &#39;deprecated_codeword_counts&#39;, &#39;total_counts&#39;, &#39;cell_area&#39;, &#39;nucleus_area&#39;, &#39;region&#39;, &#39;z_level&#39;, &#39;nucleus_count&#39;, &#39;cell_labels&#39;
    var: &#39;gene_ids&#39;, &#39;feature_types&#39;, &#39;genome&#39;
    uns: &#39;spatialdata_attrs&#39;
    obsm: &#39;spatial&#39;
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">normalize_total</span><span class="p">(</span><span class="n">sdata_crop</span><span class="o">.</span><span class="n">tables</span><span class="p">[</span><span class="s2">&quot;table&quot;</span><span class="p">],</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">log1p</span><span class="p">(</span><span class="n">sdata_crop</span><span class="o">.</span><span class="n">tables</span><span class="p">[</span><span class="s2">&quot;table&quot;</span><span class="p">])</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">pca</span><span class="p">(</span><span class="n">sdata_crop</span><span class="o">.</span><span class="n">tables</span><span class="p">[</span><span class="s2">&quot;table&quot;</span><span class="p">])</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">neighbors</span><span class="p">(</span><span class="n">sdata_crop</span><span class="o">.</span><span class="n">tables</span><span class="p">[</span><span class="s2">&quot;table&quot;</span><span class="p">])</span>
<span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">sdata_crop</span><span class="o">.</span><span class="n">tables</span><span class="p">[</span><span class="s2">&quot;table&quot;</span><span class="p">])</span>
<span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">leiden</span><span class="p">(</span><span class="n">sdata_crop</span><span class="o">.</span><span class="n">tables</span><span class="p">[</span><span class="s2">&quot;table&quot;</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/Users/tim.treis/anaconda3/envs/spatialdata-workshop/lib/python3.11/site-packages/scanpy/preprocessing/_normalization.py:235: UserWarning: Some cells have zero counts
  warn(UserWarning(&quot;Some cells have zero counts&quot;))
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span>
    <span class="n">sdata_crop</span><span class="o">.</span><span class="n">tables</span><span class="p">[</span><span class="s2">&quot;table&quot;</span><span class="p">],</span>
    <span class="n">color</span><span class="o">=</span><span class="p">[</span>
        <span class="s2">&quot;total_counts&quot;</span><span class="p">,</span>
        <span class="s2">&quot;leiden&quot;</span><span class="p">,</span>
    <span class="p">],</span>
    <span class="n">wspace</span><span class="o">=</span><span class="mf">0.4</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../../../_images/44a935f59a56352ddac9b784d065fd9ff48005d468a4c9404070dbdd3733793c.png" src="../../../_images/44a935f59a56352ddac9b784d065fd9ff48005d468a4c9404070dbdd3733793c.png" />
</div>
</div>
</section>
<section id="investigate-the-proposed-leiden-clusters">
<h2>Investigate the proposed Leiden clusters<a class="headerlink" href="#investigate-the-proposed-leiden-clusters" title="Link to this heading">#</a></h2>
<p>The Leiden clustering essentially clusters the data into an arbitrary amount of clusters based on similarity of the contained <code class="docutils literal notranslate"><span class="pre">AnnData</span></code> table. At this stage, it is recommended to directly use prior bio knowledge to annotate these clusters. We will do this in a fairly basic manner here, using marker genes for lung tissue from <a class="reference external" href="https://cellxgene.cziscience.com/gene-expression">https://cellxgene.cziscience.com/gene-expression</a>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># fmt: off</span>
<span class="n">HLCA_cell_type_signatures</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s2">&quot;alveolar_type1_fibroblasts&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;ADH1B&quot;</span><span class="p">,</span> <span class="s2">&quot;DCN&quot;</span><span class="p">,</span> <span class="s2">&quot;LUM&quot;</span><span class="p">,</span> <span class="s2">&quot;FBLN1&quot;</span><span class="p">,</span> <span class="s2">&quot;A2M&quot;</span><span class="p">,</span> <span class="s2">&quot;C1S&quot;</span><span class="p">,</span> <span class="s2">&quot;SCN7A&quot;</span><span class="p">,</span> <span class="s2">&quot;C7_ENSG00000112936&quot;</span><span class="p">,</span> <span class="s2">&quot;GPC3&quot;</span><span class="p">,</span> <span class="s2">&quot;CFD&quot;</span><span class="p">,</span> <span class="s2">&quot;FMO2&quot;</span><span class="p">,</span> <span class="s2">&quot;GPX3&quot;</span><span class="p">,</span> <span class="s2">&quot;RARRES2&quot;</span><span class="p">,</span> <span class="s2">&quot;C1R&quot;</span><span class="p">,</span> <span class="s2">&quot;PLEKHH2&quot;</span><span class="p">,</span> <span class="s2">&quot;FBLN5&quot;</span><span class="p">,</span> <span class="s2">&quot;FN1&quot;</span><span class="p">,</span> <span class="s2">&quot;PRELP&quot;</span><span class="p">,</span> <span class="s2">&quot;SELENOP&quot;</span><span class="p">,</span> <span class="s2">&quot;SERPING1&quot;</span><span class="p">,</span> <span class="s2">&quot;FHL1&quot;</span><span class="p">,</span> <span class="s2">&quot;NBL1&quot;</span><span class="p">,</span> <span class="s2">&quot;SLIT2&quot;</span><span class="p">,</span> <span class="s2">&quot;MFAP4&quot;</span><span class="p">,</span> <span class="s2">&quot;COL6A2&quot;</span><span class="p">,],</span>
    <span class="s2">&quot;alveolar_type2_fibroblast&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;DCN&quot;</span><span class="p">,</span> <span class="s2">&quot;FBLN1&quot;</span><span class="p">,</span> <span class="s2">&quot;C1S&quot;</span><span class="p">,</span> <span class="s2">&quot;IGFBP6&quot;</span><span class="p">,</span> <span class="s2">&quot;ADH1B&quot;</span><span class="p">,</span> <span class="s2">&quot;CFD&quot;</span><span class="p">,</span> <span class="s2">&quot;LUM&quot;</span><span class="p">,</span> <span class="s2">&quot;C1R&quot;</span><span class="p">,</span> <span class="s2">&quot;CCDC80&quot;</span><span class="p">,</span> <span class="s2">&quot;C3_ENSG00000125730&quot;</span><span class="p">,</span> <span class="s2">&quot;SFRP2&quot;</span><span class="p">,</span> <span class="s2">&quot;COL6A2&quot;</span><span class="p">,</span> <span class="s2">&quot;NNMT&quot;</span><span class="p">,</span> <span class="s2">&quot;MGP&quot;</span><span class="p">,</span> <span class="s2">&quot;EFEMP1&quot;</span><span class="p">,</span> <span class="s2">&quot;FBN1&quot;</span><span class="p">,</span> <span class="s2">&quot;GSN&quot;</span><span class="p">,</span> <span class="s2">&quot;FSTL1&quot;</span><span class="p">,</span> <span class="s2">&quot;PLAC9&quot;</span><span class="p">,</span> <span class="s2">&quot;C7_ENSG00000112936&quot;</span><span class="p">,</span> <span class="s2">&quot;ABI3BP&quot;</span><span class="p">,</span> <span class="s2">&quot;SERPING1&quot;</span><span class="p">,</span> <span class="s2">&quot;COL1A2&quot;</span><span class="p">,</span> <span class="s2">&quot;MFAP5&quot;</span><span class="p">,</span> <span class="s2">&quot;RARRES1&quot;</span><span class="p">,],</span>
    <span class="s2">&quot;endothelial_cell&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;CLDN5&quot;</span><span class="p">,</span> <span class="s2">&quot;PECAM1&quot;</span><span class="p">,</span> <span class="s2">&quot;EPAS1&quot;</span><span class="p">,</span> <span class="s2">&quot;EGFL7&quot;</span><span class="p">,</span> <span class="s2">&quot;RAMP2&quot;</span><span class="p">,</span> <span class="s2">&quot;CALCRL&quot;</span><span class="p">,</span> <span class="s2">&quot;GNG11&quot;</span><span class="p">,</span> <span class="s2">&quot;VWF&quot;</span><span class="p">,</span> <span class="s2">&quot;AQP1&quot;</span><span class="p">,</span> <span class="s2">&quot;HYAL2&quot;</span><span class="p">,</span> <span class="s2">&quot;TM4SF1&quot;</span><span class="p">,</span> <span class="s2">&quot;ENG&quot;</span><span class="p">,</span> <span class="s2">&quot;CLEC14A&quot;</span><span class="p">,</span> <span class="s2">&quot;CAVIN2&quot;</span><span class="p">,</span> <span class="s2">&quot;CDH5&quot;</span><span class="p">,</span> <span class="s2">&quot;PTPRB&quot;</span><span class="p">,</span> <span class="s2">&quot;RAMP3&quot;</span><span class="p">,</span> <span class="s2">&quot;TFPI&quot;</span><span class="p">,</span> <span class="s2">&quot;ARHGAP29&quot;</span><span class="p">,</span> <span class="s2">&quot;CD93&quot;</span><span class="p">,</span> <span class="s2">&quot;SLCO2A1&quot;</span><span class="p">,</span> <span class="s2">&quot;SPARCL1&quot;</span><span class="p">,</span> <span class="s2">&quot;LDB2&quot;</span><span class="p">,</span> <span class="s2">&quot;CAV1&quot;</span><span class="p">,</span> <span class="s2">&quot;PCAT19&quot;</span><span class="p">,],</span>
    <span class="s2">&quot;alveolar_macrophage&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;C1QA&quot;</span><span class="p">,</span> <span class="s2">&quot;APOC1&quot;</span><span class="p">,</span> <span class="s2">&quot;C1QB&quot;</span><span class="p">,</span> <span class="s2">&quot;ACP5&quot;</span><span class="p">,</span> <span class="s2">&quot;MARCO&quot;</span><span class="p">,</span> <span class="s2">&quot;C1QC&quot;</span><span class="p">,</span> <span class="s2">&quot;GRN&quot;</span><span class="p">,</span> <span class="s2">&quot;CTSS&quot;</span><span class="p">,</span> <span class="s2">&quot;FABP4&quot;</span><span class="p">,</span> <span class="s2">&quot;FBP1&quot;</span><span class="p">,</span> <span class="s2">&quot;PSAP&quot;</span><span class="p">,</span> <span class="s2">&quot;FTL&quot;</span><span class="p">,</span> <span class="s2">&quot;MS4A7&quot;</span><span class="p">,</span> <span class="s2">&quot;MRC1&quot;</span><span class="p">,</span> <span class="s2">&quot;MCEMP1&quot;</span><span class="p">,</span> <span class="s2">&quot;CD68&quot;</span><span class="p">,</span> <span class="s2">&quot;VSIG4&quot;</span><span class="p">,</span> <span class="s2">&quot;MSR1&quot;</span><span class="p">,</span> <span class="s2">&quot;OLR1&quot;</span><span class="p">,</span> <span class="s2">&quot;APOE&quot;</span><span class="p">,</span> <span class="s2">&quot;CTSD&quot;</span><span class="p">,</span> <span class="s2">&quot;ALOX5AP&quot;</span><span class="p">,</span> <span class="s2">&quot;CTSC&quot;</span><span class="p">,</span> <span class="s2">&quot;GPNMB&quot;</span><span class="p">,</span> <span class="s2">&quot;CAPG&quot;</span><span class="p">,],</span>
    <span class="s2">&quot;lung_interstitial_macrophage&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;C1QB&quot;</span><span class="p">,</span> <span class="s2">&quot;C1QA&quot;</span><span class="p">,</span> <span class="s2">&quot;C1QC&quot;</span><span class="p">,</span> <span class="s2">&quot;CTSS&quot;</span><span class="p">,</span> <span class="s2">&quot;CD68&quot;</span><span class="p">,</span> <span class="s2">&quot;CD14&quot;</span><span class="p">,</span> <span class="s2">&quot;CTSZ&quot;</span><span class="p">,</span> <span class="s2">&quot;MS4A7&quot;</span><span class="p">,</span> <span class="s2">&quot;CTSD&quot;</span><span class="p">,</span> <span class="s2">&quot;PSAP&quot;</span><span class="p">,</span> <span class="s2">&quot;IFI30&quot;</span><span class="p">,</span> <span class="s2">&quot;CD163&quot;</span><span class="p">,</span> <span class="s2">&quot;MS4A6A&quot;</span><span class="p">,</span> <span class="s2">&quot;GRN&quot;</span><span class="p">,</span> <span class="s2">&quot;MRC1&quot;</span><span class="p">,</span> <span class="s2">&quot;APOE&quot;</span><span class="p">,</span> <span class="s2">&quot;CYBB&quot;</span><span class="p">,</span> <span class="s2">&quot;MS4A4A&quot;</span><span class="p">,</span> <span class="s2">&quot;NPC2&quot;</span><span class="p">,</span> <span class="s2">&quot;VSIG4&quot;</span><span class="p">,</span> <span class="s2">&quot;FCGR3A&quot;</span><span class="p">,</span> <span class="s2">&quot;FTL&quot;</span><span class="p">,</span> <span class="s2">&quot;CTSC&quot;</span><span class="p">,</span> <span class="s2">&quot;TYMP&quot;</span><span class="p">,</span> <span class="s2">&quot;GPNMB&quot;</span><span class="p">,],</span>
    <span class="s2">&quot;fibroblast&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;DCN&quot;</span><span class="p">,</span> <span class="s2">&quot;LUM&quot;</span><span class="p">,</span> <span class="s2">&quot;FBLN1&quot;</span><span class="p">,</span> <span class="s2">&quot;COL6A2&quot;</span><span class="p">,</span> <span class="s2">&quot;COL1A2&quot;</span><span class="p">,</span> <span class="s2">&quot;ADH1B&quot;</span><span class="p">,</span> <span class="s2">&quot;COL6A3&quot;</span><span class="p">,</span> <span class="s2">&quot;C1S&quot;</span><span class="p">,</span> <span class="s2">&quot;C7_ENSG00000112936&quot;</span><span class="p">,</span> <span class="s2">&quot;COL6A1&quot;</span><span class="p">,</span> <span class="s2">&quot;C1R&quot;</span><span class="p">,</span> <span class="s2">&quot;FN1&quot;</span><span class="p">,</span> <span class="s2">&quot;MFAP4&quot;</span><span class="p">,</span> <span class="s2">&quot;GPC3&quot;</span><span class="p">,</span> <span class="s2">&quot;COL3A1&quot;</span><span class="p">,</span> <span class="s2">&quot;CDH11&quot;</span><span class="p">,</span> <span class="s2">&quot;RARRES2&quot;</span><span class="p">,</span> <span class="s2">&quot;PCOLCE&quot;</span><span class="p">,</span> <span class="s2">&quot;COL1A1&quot;</span><span class="p">,</span> <span class="s2">&quot;NBL1&quot;</span><span class="p">,</span> <span class="s2">&quot;TCF21&quot;</span><span class="p">,</span> <span class="s2">&quot;OLFML3&quot;</span><span class="p">,</span> <span class="s2">&quot;CALD1&quot;</span><span class="p">,</span> <span class="s2">&quot;CFD&quot;</span><span class="p">,</span> <span class="s2">&quot;COL5A2&quot;</span><span class="p">,],</span>
    <span class="s2">&quot;smooth_muscle_cell&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;ACTA2&quot;</span><span class="p">,</span> <span class="s2">&quot;TAGLN&quot;</span><span class="p">,</span> <span class="s2">&quot;TPM2&quot;</span><span class="p">,</span> <span class="s2">&quot;MYL9&quot;</span><span class="p">,</span> <span class="s2">&quot;CALD1&quot;</span><span class="p">,</span> <span class="s2">&quot;MYH11&quot;</span><span class="p">,</span> <span class="s2">&quot;IGFBP7&quot;</span><span class="p">,</span> <span class="s2">&quot;TPM1&quot;</span><span class="p">,</span> <span class="s2">&quot;SPARCL1&quot;</span><span class="p">,</span> <span class="s2">&quot;C11orf96&quot;</span><span class="p">,</span> <span class="s2">&quot;PPP1R14A&quot;</span><span class="p">,</span> <span class="s2">&quot;SELENOM&quot;</span><span class="p">,</span> <span class="s2">&quot;DSTN&quot;</span><span class="p">,</span> <span class="s2">&quot;MYLK&quot;</span><span class="p">,</span> <span class="s2">&quot;ACTG2&quot;</span><span class="p">,</span> <span class="s2">&quot;ADIRF&quot;</span><span class="p">,</span> <span class="s2">&quot;BGN&quot;</span><span class="p">,</span> <span class="s2">&quot;SOD3&quot;</span><span class="p">,</span> <span class="s2">&quot;MFGE8&quot;</span><span class="p">,</span> <span class="s2">&quot;COL6A2&quot;</span><span class="p">,</span> <span class="s2">&quot;LMOD1&quot;</span><span class="p">,</span> <span class="s2">&quot;TINAGL1&quot;</span><span class="p">,</span> <span class="s2">&quot;CAVIN3&quot;</span><span class="p">,</span> <span class="s2">&quot;DES&quot;</span><span class="p">,</span> <span class="s2">&quot;CNN1&quot;</span><span class="p">,],</span>
    <span class="s2">&quot;mast_cell&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;CPA3&quot;</span><span class="p">,</span> <span class="s2">&quot;TPSAB1&quot;</span><span class="p">,</span> <span class="s2">&quot;TPSB2&quot;</span><span class="p">,</span> <span class="s2">&quot;HPGDS&quot;</span><span class="p">,</span> <span class="s2">&quot;MS4A2&quot;</span><span class="p">,</span> <span class="s2">&quot;KIT&quot;</span><span class="p">,</span> <span class="s2">&quot;GATA2&quot;</span><span class="p">,</span> <span class="s2">&quot;VWA5A&quot;</span><span class="p">,</span> <span class="s2">&quot;SLC18A2&quot;</span><span class="p">,</span> <span class="s2">&quot;IL1RL1&quot;</span><span class="p">,</span> <span class="s2">&quot;HPGD&quot;</span><span class="p">,</span> <span class="s2">&quot;RGS13&quot;</span><span class="p">,</span> <span class="s2">&quot;SAMSN1&quot;</span><span class="p">,</span> <span class="s2">&quot;RHEX&quot;</span><span class="p">,</span> <span class="s2">&quot;RGS1&quot;</span><span class="p">,</span> <span class="s2">&quot;HDC&quot;</span><span class="p">,</span> <span class="s2">&quot;AREG&quot;</span><span class="p">,</span> <span class="s2">&quot;LTC4S&quot;</span><span class="p">,</span> <span class="s2">&quot;FOSB&quot;</span><span class="p">,</span> <span class="s2">&quot;PTGS1&quot;</span><span class="p">,</span> <span class="s2">&quot;CAPG&quot;</span><span class="p">,</span> <span class="s2">&quot;CD69&quot;</span><span class="p">,</span> <span class="s2">&quot;ARHGEF6&quot;</span><span class="p">,</span> <span class="s2">&quot;PTGS2&quot;</span><span class="p">,</span> <span class="s2">&quot;CD44&quot;</span><span class="p">,],</span>
    <span class="s2">&quot;dendritic_cell&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;HLA-DQB1&quot;</span><span class="p">,</span> <span class="s2">&quot;CD74&quot;</span><span class="p">,</span> <span class="s2">&quot;GPR183&quot;</span><span class="p">,</span> <span class="s2">&quot;MS4A6A&quot;</span><span class="p">,</span> <span class="s2">&quot;HLA-DPB1&quot;</span><span class="p">,</span> <span class="s2">&quot;HLA-DRB5&quot;</span><span class="p">,</span> <span class="s2">&quot;HLA-DQA1&quot;</span><span class="p">,</span> <span class="s2">&quot;HLA-DPA1&quot;</span><span class="p">,</span> <span class="s2">&quot;CLEC10A&quot;</span><span class="p">,</span> <span class="s2">&quot;CD83&quot;</span><span class="p">,</span> <span class="s2">&quot;CST3&quot;</span><span class="p">,</span> <span class="s2">&quot;CPVL&quot;</span><span class="p">,</span> <span class="s2">&quot;HLA-DMA&quot;</span><span class="p">,</span> <span class="s2">&quot;FGL2&quot;</span><span class="p">,</span> <span class="s2">&quot;CD1C&quot;</span><span class="p">,</span> <span class="s2">&quot;CD86&quot;</span><span class="p">,</span> <span class="s2">&quot;RNASE6&quot;</span><span class="p">,</span> <span class="s2">&quot;FCER1A&quot;</span><span class="p">,</span> <span class="s2">&quot;CSF2RA&quot;</span><span class="p">,</span> <span class="s2">&quot;TYMP&quot;</span><span class="p">,</span> <span class="s2">&quot;RGS1&quot;</span><span class="p">,</span> <span class="s2">&quot;RGS10&quot;</span><span class="p">,</span> <span class="s2">&quot;C1orf162&quot;</span><span class="p">,</span> <span class="s2">&quot;GRN&quot;</span><span class="p">,</span> <span class="s2">&quot;AREG&quot;</span><span class="p">,],</span>
<span class="p">}</span>
<span class="c1"># fmt: on</span>
</pre></div>
</div>
</div>
</div>
<p>It is expected that we find only a subset of these genes since Xenium itself only profiles a small portion of the entire transcriptome.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span><span class="w"> </span><span class="nf">find_signature_markers</span><span class="p">(</span><span class="n">signatures</span><span class="p">,</span> <span class="n">data_columns</span><span class="p">):</span>
    <span class="n">found_markers</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="n">cell_type</span><span class="p">,</span> <span class="n">genes</span> <span class="ow">in</span> <span class="n">signatures</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="n">found_markers</span><span class="p">[</span><span class="n">cell_type</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">gene</span> <span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">genes</span> <span class="k">if</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">data_columns</span><span class="p">]</span>
        <span class="nb">print</span><span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;Found </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">found_markers</span><span class="p">[</span><span class="n">cell_type</span><span class="p">])</span><span class="si">}</span><span class="s2"> genes of the </span><span class="si">{</span><span class="n">cell_type</span><span class="si">}</span><span class="s2"> signature: </span><span class="si">{</span><span class="n">found_markers</span><span class="p">[</span><span class="n">cell_type</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span>
        <span class="p">)</span>
    <span class="k">return</span> <span class="n">found_markers</span>


<span class="n">found_markers</span> <span class="o">=</span> <span class="n">find_signature_markers</span><span class="p">(</span>
    <span class="n">HLCA_cell_type_signatures</span><span class="p">,</span> <span class="n">sdata_crop</span><span class="o">.</span><span class="n">tables</span><span class="p">[</span><span class="s2">&quot;table&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">to_df</span><span class="p">()</span><span class="o">.</span><span class="n">columns</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Found 2 genes of the alveolar_type1_fibroblasts signature: [&#39;FBLN1&#39;, &#39;GPC3&#39;]
Found 5 genes of the alveolar_type2_fibroblast signature: [&#39;FBLN1&#39;, &#39;SFRP2&#39;, &#39;FBN1&#39;, &#39;PLAC9&#39;, &#39;MFAP5&#39;]
Found 10 genes of the endothelial_cell signature: [&#39;PECAM1&#39;, &#39;EGFL7&#39;, &#39;RAMP2&#39;, &#39;GNG11&#39;, &#39;VWF&#39;, &#39;CLEC14A&#39;, &#39;CAVIN2&#39;, &#39;TFPI&#39;, &#39;CD93&#39;, &#39;CAV1&#39;]
Found 5 genes of the alveolar_macrophage signature: [&#39;MARCO&#39;, &#39;MRC1&#39;, &#39;MCEMP1&#39;, &#39;CD68&#39;, &#39;VSIG4&#39;]
Found 8 genes of the lung_interstitial_macrophage signature: [&#39;CD68&#39;, &#39;CD14&#39;, &#39;CD163&#39;, &#39;MS4A6A&#39;, &#39;MRC1&#39;, &#39;MS4A4A&#39;, &#39;VSIG4&#39;, &#39;FCGR3A&#39;]
Found 4 genes of the fibroblast signature: [&#39;FBLN1&#39;, &#39;GPC3&#39;, &#39;PCOLCE&#39;, &#39;COL5A2&#39;]
Found 6 genes of the smooth_muscle_cell signature: [&#39;ACTA2&#39;, &#39;MYH11&#39;, &#39;MYLK&#39;, &#39;ACTG2&#39;, &#39;DES&#39;, &#39;CNN1&#39;]
Found 9 genes of the mast_cell signature: [&#39;CPA3&#39;, &#39;HPGDS&#39;, &#39;MS4A2&#39;, &#39;KIT&#39;, &#39;GATA2&#39;, &#39;VWA5A&#39;, &#39;SLC18A2&#39;, &#39;IL1RL1&#39;, &#39;CD69&#39;]
Found 10 genes of the dendritic_cell signature: [&#39;GPR183&#39;, &#39;MS4A6A&#39;, &#39;CLEC10A&#39;, &#39;CD83&#39;, &#39;FGL2&#39;, &#39;CD1C&#39;, &#39;CD86&#39;, &#39;FCER1A&#39;, &#39;CSF2RA&#39;, &#39;C1orf162&#39;]
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span><span class="w"> </span><span class="nf">rank_and_normalize_gexp</span><span class="p">(</span><span class="n">series</span><span class="p">):</span>
    <span class="n">ranked</span> <span class="o">=</span> <span class="n">series</span><span class="o">.</span><span class="n">rank</span><span class="p">(</span><span class="n">ascending</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s2">&quot;min&quot;</span><span class="p">)</span>
    <span class="n">min_rank</span> <span class="o">=</span> <span class="n">ranked</span><span class="o">.</span><span class="n">min</span><span class="p">()</span>
    <span class="n">max_rank</span> <span class="o">=</span> <span class="n">ranked</span><span class="o">.</span><span class="n">max</span><span class="p">()</span>

    <span class="k">return</span> <span class="n">ranked</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="p">(</span><span class="n">x</span> <span class="o">-</span> <span class="n">min_rank</span><span class="p">)</span> <span class="o">/</span> <span class="p">(</span><span class="n">max_rank</span> <span class="o">-</span> <span class="n">min_rank</span><span class="p">))</span>


<span class="k">def</span><span class="w"> </span><span class="nf">make_score_from_signature</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">signature</span><span class="p">,</span> <span class="n">zero_as_none</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>

    <span class="k">return</span> <span class="p">(</span>
        <span class="n">adata</span><span class="o">.</span><span class="n">to_df</span><span class="p">()[</span><span class="n">signature</span><span class="p">]</span>
        <span class="o">.</span><span class="n">rank</span><span class="p">(</span><span class="n">ascending</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s2">&quot;min&quot;</span><span class="p">)</span>  <span class="c1"># Sort genes by rank, ties = equal rank</span>
        <span class="o">.</span><span class="n">apply</span><span class="p">(</span>
            <span class="k">lambda</span> <span class="n">col</span><span class="p">:</span> <span class="n">rank_and_normalize_gexp</span><span class="p">(</span><span class="n">col</span><span class="p">)</span>
        <span class="p">)</span>  <span class="c1"># Normalize the ranks to [0, 1]</span>
        <span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>  <span class="c1"># Sum the normalized ranks</span>
        <span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">row</span><span class="p">:</span> <span class="kc">None</span> <span class="k">if</span> <span class="n">zero_as_none</span> <span class="ow">and</span> <span class="n">row</span> <span class="o">==</span> <span class="mi">0</span> <span class="k">else</span> <span class="n">row</span><span class="p">)</span>
    <span class="p">)</span>


<span class="k">def</span><span class="w"> </span><span class="nf">compute_signature_scores</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">found_markers</span><span class="p">):</span>
    <span class="k">for</span> <span class="n">cell_type</span><span class="p">,</span> <span class="n">markers</span> <span class="ow">in</span> <span class="n">found_markers</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">cell_type</span> <span class="o">+</span> <span class="s2">&quot;_score&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">make_score_from_signature</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">markers</span><span class="p">)</span>
    <span class="n">score_colnames</span> <span class="o">=</span> <span class="p">[</span><span class="n">cell_type</span> <span class="o">+</span> <span class="s2">&quot;_score&quot;</span> <span class="k">for</span> <span class="n">cell_type</span> <span class="ow">in</span> <span class="n">found_markers</span><span class="o">.</span><span class="n">keys</span><span class="p">()]</span>
    <span class="k">return</span> <span class="n">score_colnames</span>


<span class="n">score_colnames</span> <span class="o">=</span> <span class="n">compute_signature_scores</span><span class="p">(</span><span class="n">sdata_crop</span><span class="o">.</span><span class="n">tables</span><span class="p">[</span><span class="s2">&quot;table&quot;</span><span class="p">],</span> <span class="n">found_markers</span><span class="p">)</span>

<span class="n">sdata_crop</span><span class="o">.</span><span class="n">tables</span><span class="p">[</span><span class="s2">&quot;table&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">score_colnames</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>alveolar_type1_fibroblasts_score</th>
      <th>alveolar_type2_fibroblast_score</th>
      <th>endothelial_cell_score</th>
      <th>alveolar_macrophage_score</th>
      <th>lung_interstitial_macrophage_score</th>
      <th>fibroblast_score</th>
      <th>smooth_muscle_cell_score</th>
      <th>mast_cell_score</th>
      <th>dendritic_cell_score</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>22691</th>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>0.951225</td>
    </tr>
    <tr>
      <th>22692</th>
      <td>0.929370</td>
      <td>1.851537</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>0.930111</td>
      <td>0.929370</td>
      <td>NaN</td>
      <td>0.987280</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>22693</th>
      <td>0.935953</td>
      <td>0.935953</td>
      <td>0.877780</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>0.935953</td>
      <td>0.922844</td>
      <td>0.970498</td>
      <td>1.909476</td>
    </tr>
    <tr>
      <th>22694</th>
      <td>0.929370</td>
      <td>0.929370</td>
      <td>0.954664</td>
      <td>NaN</td>
      <td>0.867665</td>
      <td>0.929370</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>0.867665</td>
    </tr>
    <tr>
      <th>22695</th>
      <td>NaN</td>
      <td>NaN</td>
      <td>0.950543</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>0.969193</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>162118</th>
      <td>NaN</td>
      <td>NaN</td>
      <td>0.988318</td>
      <td>NaN</td>
      <td>1.995463</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>0.999318</td>
    </tr>
    <tr>
      <th>162119</th>
      <td>NaN</td>
      <td>NaN</td>
      <td>0.960209</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>0.991905</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>162120</th>
      <td>NaN</td>
      <td>NaN</td>
      <td>0.772520</td>
      <td>0.956918</td>
      <td>1.878133</td>
      <td>NaN</td>
      <td>0.965633</td>
      <td>NaN</td>
      <td>0.921216</td>
    </tr>
    <tr>
      <th>162121</th>
      <td>NaN</td>
      <td>0.999970</td>
      <td>NaN</td>
      <td>0.999644</td>
      <td>0.999644</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>162122</th>
      <td>NaN</td>
      <td>NaN</td>
      <td>0.996709</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
  </tbody>
</table>
<p>33727 rows × 9 columns</p>
</div></div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span>
    <span class="n">sdata_crop</span><span class="o">.</span><span class="n">tables</span><span class="p">[</span><span class="s2">&quot;table&quot;</span><span class="p">],</span>
    <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;leiden&quot;</span><span class="p">]</span> <span class="o">+</span> <span class="n">score_colnames</span><span class="p">,</span>
    <span class="n">wspace</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../../../_images/f4a756207fbe31716ed78f1e01fabe077a8ba3f6385bea482c9586f3aed79b57.png" src="../../../_images/f4a756207fbe31716ed78f1e01fabe077a8ba3f6385bea482c9586f3aed79b57.png" />
</div>
</div>
<p>While we were able to identify some of the celltypes represented by the clusters, in particular fibroblasts, endothelial cells, muscle cells and mast cells, some clusters were either not amongst these cell types or were grouped into the same cluster, for example the dendritic cells and macrophages which form leiden cluster #3. We can already appreciate the following mapping:</p>
<ul class="simple">
<li><p>Cluster 0: ???</p></li>
<li><p>Cluster 1: Smooth muscle cells</p></li>
<li><p>Cluster 2: ???</p></li>
<li><p>Cluster 3: Macrophages &amp; Dendritic cells</p></li>
<li><p>Cluster 4: Fibroblasts</p></li>
<li><p>Cluster 5: ???</p></li>
<li><p>Cluster 6: Endothelial cells</p></li>
<li><p>Cluster 7: ???</p></li>
<li><p>Cluster 8: ???</p></li>
<li><p>Cluster 9: ???</p></li>
<li><p>Cluster 10: ???</p></li>
<li><p>Cluster 11: Mast cells</p></li>
</ul>
<p>Since the clusters 0, 2, and 5, represent a large portion of our cells, we’ll try to identify these cells by more closely looking into their expression values.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_025</span> <span class="o">=</span> <span class="n">sdata_crop</span><span class="o">.</span><span class="n">tables</span><span class="p">[</span><span class="s2">&quot;table&quot;</span><span class="p">][</span>
    <span class="n">sdata_crop</span><span class="o">.</span><span class="n">tables</span><span class="p">[</span><span class="s2">&quot;table&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;leiden&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">([</span><span class="s2">&quot;0&quot;</span><span class="p">,</span> <span class="s2">&quot;2&quot;</span><span class="p">,</span> <span class="s2">&quot;5&quot;</span><span class="p">])</span>
<span class="p">]</span>
<span class="n">adata_025</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>View of AnnData object with n_obs × n_vars = 15650 × 377
    obs: &#39;cell_id&#39;, &#39;transcript_counts&#39;, &#39;control_probe_counts&#39;, &#39;control_codeword_counts&#39;, &#39;unassigned_codeword_counts&#39;, &#39;deprecated_codeword_counts&#39;, &#39;total_counts&#39;, &#39;cell_area&#39;, &#39;nucleus_area&#39;, &#39;region&#39;, &#39;z_level&#39;, &#39;nucleus_count&#39;, &#39;cell_labels&#39;, &#39;leiden&#39;, &#39;alveolar_type1_fibroblasts_score&#39;, &#39;alveolar_type2_fibroblast_score&#39;, &#39;endothelial_cell_score&#39;, &#39;alveolar_macrophage_score&#39;, &#39;lung_interstitial_macrophage_score&#39;, &#39;fibroblast_score&#39;, &#39;smooth_muscle_cell_score&#39;, &#39;mast_cell_score&#39;, &#39;dendritic_cell_score&#39;
    var: &#39;gene_ids&#39;, &#39;feature_types&#39;, &#39;genome&#39;
    uns: &#39;spatialdata_attrs&#39;, &#39;log1p&#39;, &#39;pca&#39;, &#39;neighbors&#39;, &#39;umap&#39;, &#39;leiden&#39;, &#39;leiden_colors&#39;
    obsm: &#39;spatial&#39;, &#39;X_pca&#39;, &#39;X_umap&#39;
    varm: &#39;PCs&#39;
    obsp: &#39;distances&#39;, &#39;connectivities&#39;
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">highest_expr_genes</span><span class="p">(</span><span class="n">adata_025</span><span class="p">,</span> <span class="n">n_top</span><span class="o">=</span><span class="mi">20</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/Users/tim.treis/anaconda3/envs/spatialdata-workshop/lib/python3.11/site-packages/scanpy/preprocessing/_normalization.py:208: UserWarning: Received a view of an AnnData. Making a copy.
  view_to_actual(adata)
</pre></div>
</div>
<img alt="../../../_images/4f57c9b93dd4c367ff5efa3abe2f78dcdba4e77dfdde82f26a78f3525dacba0d.png" src="../../../_images/4f57c9b93dd4c367ff5efa3abe2f78dcdba4e77dfdde82f26a78f3525dacba0d.png" />
</div>
</div>
<p>We can extract these genes and then use the HLCA explorer by CELLxGENE to try to find out which cells the clusters could be. For this, we’ll paste the “signature” into their GUI.</p>
<blockquote>
<div><p>MALL, CYP2B6, TCIM, CAPN8, TFPI, EPCAM, GPRC5A, CYP4B1, TMC5, DST, KRT7, CFTR, EHF, NTN4, CFB, STEAP4, MDM2, BASP1, SFTA2, MLPH</p>
</div></blockquote>
<p>From there we can extract several cell types which highly express these genes:</p>
<p><img alt="image.png" src="../../../_images/nb4_cellxgene.png" /></p>
<p>We hand-picked a few of the signatures and will repeat the above steps:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># fmt: off</span>
<span class="n">proposed_celltype_signatures</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s2">&quot;pneumocyte&quot;</span><span class="p">:</span> <span class="p">[</span>
        <span class="s2">&quot;SFTPB&quot;</span><span class="p">,</span> <span class="s2">&quot;NAPSA&quot;</span><span class="p">,</span> <span class="s2">&quot;SFTPA1&quot;</span><span class="p">,</span> <span class="s2">&quot;SLC34A2&quot;</span><span class="p">,</span> <span class="s2">&quot;SFTPA2&quot;</span><span class="p">,</span> <span class="s2">&quot;SFTPC&quot;</span><span class="p">,</span> <span class="s2">&quot;SFTPD&quot;</span><span class="p">,</span> <span class="s2">&quot;PEBP4&quot;</span><span class="p">,</span> <span class="s2">&quot;CTSH&quot;</span><span class="p">,</span> <span class="s2">&quot;HOPX&quot;</span><span class="p">,</span> <span class="s2">&quot;SFTA3_ENSG00000229415&quot;</span><span class="p">,</span> <span class="s2">&quot;LAMP3&quot;</span><span class="p">,</span> <span class="s2">&quot;ABCA3&quot;</span><span class="p">,</span> <span class="s2">&quot;PGC&quot;</span><span class="p">,</span> <span class="s2">&quot;SLC39A8&quot;</span><span class="p">,</span> <span class="s2">&quot;CLDN18&quot;</span><span class="p">,</span> <span class="s2">&quot;MUC1&quot;</span><span class="p">,</span> <span class="s2">&quot;C4BPA&quot;</span><span class="p">,</span> <span class="s2">&quot;DHCR24&quot;</span><span class="p">,</span> <span class="s2">&quot;SDR16C5&quot;</span><span class="p">,</span> <span class="s2">&quot;WIF1&quot;</span><span class="p">,</span> <span class="s2">&quot;C16orf89&quot;</span><span class="p">,</span> <span class="s2">&quot;SDC4&quot;</span><span class="p">,</span> <span class="s2">&quot;SFTA2&quot;</span><span class="p">,</span> <span class="s2">&quot;ATP11A&quot;</span><span class="p">,],</span>
    <span class="s2">&quot;club_cell&quot;</span><span class="p">:</span> <span class="p">[</span>
        <span class="s2">&quot;PIGR&quot;</span><span class="p">,</span> <span class="s2">&quot;CXCL17&quot;</span><span class="p">,</span> <span class="s2">&quot;SCGB3A1&quot;</span><span class="p">,</span> <span class="s2">&quot;SCGB1A1&quot;</span><span class="p">,</span> <span class="s2">&quot;SFTPB&quot;</span><span class="p">,</span> <span class="s2">&quot;SLPI&quot;</span><span class="p">,</span> <span class="s2">&quot;WFDC2&quot;</span><span class="p">,</span> <span class="s2">&quot;BPIFB1&quot;</span><span class="p">,</span> <span class="s2">&quot;CP&quot;</span><span class="p">,</span> <span class="s2">&quot;CEACAM6&quot;</span><span class="p">,</span> <span class="s2">&quot;C16orf89&quot;</span><span class="p">,</span> <span class="s2">&quot;KLK11&quot;</span><span class="p">,</span> <span class="s2">&quot;SCGB3A2&quot;</span><span class="p">,</span> <span class="s2">&quot;CYP4B1&quot;</span><span class="p">,</span> <span class="s2">&quot;CYP2F1&quot;</span><span class="p">,</span> <span class="s2">&quot;LCN2&quot;</span><span class="p">,</span> <span class="s2">&quot;AGR3&quot;</span><span class="p">,</span> <span class="s2">&quot;STEAP4&quot;</span><span class="p">,</span> <span class="s2">&quot;NEDD4L&quot;</span><span class="p">,],</span>
    <span class="s2">&quot;basal_cell&quot;</span><span class="p">:</span> <span class="p">[</span>
        <span class="s2">&quot;S100A2&quot;</span><span class="p">,</span> <span class="s2">&quot;KRT17&quot;</span><span class="p">,</span> <span class="s2">&quot;AQP3&quot;</span><span class="p">,</span> <span class="s2">&quot;KRT5&quot;</span><span class="p">,</span> <span class="s2">&quot;KRT15&quot;</span><span class="p">,</span> <span class="s2">&quot;TACSTD2&quot;</span><span class="p">,</span> <span class="s2">&quot;KRT19&quot;</span><span class="p">,</span> <span class="s2">&quot;F3_ENSG00000117525&quot;</span><span class="p">,</span> <span class="s2">&quot;SERPINB3&quot;</span><span class="p">,</span> <span class="s2">&quot;ALDH3A1&quot;</span><span class="p">,</span> <span class="s2">&quot;CLDN1&quot;</span><span class="p">,</span> <span class="s2">&quot;CXCL1&quot;</span><span class="p">,</span> <span class="s2">&quot;MIR205HG&quot;</span><span class="p">,],</span>
<span class="p">}</span>
<span class="c1"># fmt: on</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">found_markers_in_unknown_cluster</span> <span class="o">=</span> <span class="n">find_signature_markers</span><span class="p">(</span>
    <span class="n">proposed_celltype_signatures</span><span class="p">,</span> <span class="n">sdata_crop</span><span class="o">.</span><span class="n">tables</span><span class="p">[</span><span class="s2">&quot;table&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">to_df</span><span class="p">()</span><span class="o">.</span><span class="n">columns</span>
<span class="p">)</span>

<span class="n">additional_score_colnames</span> <span class="o">=</span> <span class="n">compute_signature_scores</span><span class="p">(</span>
    <span class="n">sdata_crop</span><span class="o">.</span><span class="n">tables</span><span class="p">[</span><span class="s2">&quot;table&quot;</span><span class="p">],</span> <span class="n">found_markers_in_unknown_cluster</span>
<span class="p">)</span>

<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span>
    <span class="n">sdata_crop</span><span class="o">.</span><span class="n">tables</span><span class="p">[</span><span class="s2">&quot;table&quot;</span><span class="p">],</span>
    <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;leiden&quot;</span><span class="p">]</span> <span class="o">+</span> <span class="n">additional_score_colnames</span><span class="p">,</span>
    <span class="n">wspace</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Found 3 genes of the pneumocyte signature: [&#39;PEBP4&#39;, &#39;LAMP3&#39;, &#39;SFTA2&#39;]
Found 5 genes of the club_cell signature: [&#39;KLK11&#39;, &#39;CYP4B1&#39;, &#39;CYP2F1&#39;, &#39;AGR3&#39;, &#39;STEAP4&#39;]
Found 2 genes of the basal_cell signature: [&#39;AQP3&#39;, &#39;SERPINB3&#39;]
</pre></div>
</div>
<img alt="../../../_images/812e5ce2f7cdcea3078a2199330d97113b1974ebcc5410f9918a543a0798fdc9.png" src="../../../_images/812e5ce2f7cdcea3078a2199330d97113b1974ebcc5410f9918a543a0798fdc9.png" />
</div>
</div>
<p>While these plots might suggest that the 3 proposed cell-types are somewhere in there, we can also see that only very few genes of those signatures are used for every annotation. This can be attributed to the limited size of the Xenium panel. One possible step to better analse these cells is to subcluster and reanalyse them. Within the scope of this tutorial, we will however not proceed with this due to time limitations. Instead, we will assume that Leiden cluster 7 represents the Club cells and label clusters 0, 2, and 5 as “Pneumocytes / Basal Cells”</p>
<section id="rename-clusters">
<h3>Rename clusters<a class="headerlink" href="#rename-clusters" title="Link to this heading">#</a></h3>
<p>Based on our bio-investigation of our Leiden clusters, we can now use this information to localise them in the tissue.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">celltype_mapping</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s2">&quot;0&quot;</span><span class="p">:</span> <span class="s2">&quot;Pneumocytes / Basal Cells&quot;</span><span class="p">,</span>
    <span class="s2">&quot;1&quot;</span><span class="p">:</span> <span class="s2">&quot;Smooth muscle cells&quot;</span><span class="p">,</span>
    <span class="s2">&quot;2&quot;</span><span class="p">:</span> <span class="s2">&quot;Pneumocytes / Basal Cells&quot;</span><span class="p">,</span>
    <span class="s2">&quot;3&quot;</span><span class="p">:</span> <span class="s2">&quot;Macrophages &amp; Dendritic cells&quot;</span><span class="p">,</span>
    <span class="s2">&quot;4&quot;</span><span class="p">:</span> <span class="s2">&quot;Fibroblasts&quot;</span><span class="p">,</span>
    <span class="s2">&quot;5&quot;</span><span class="p">:</span> <span class="s2">&quot;Pneumocytes / Basal Cells&quot;</span><span class="p">,</span>
    <span class="s2">&quot;6&quot;</span><span class="p">:</span> <span class="s2">&quot;Endothelial cells&quot;</span><span class="p">,</span>
    <span class="s2">&quot;7&quot;</span><span class="p">:</span> <span class="s2">&quot;Club cells&quot;</span><span class="p">,</span>
    <span class="s2">&quot;11&quot;</span><span class="p">:</span> <span class="s2">&quot;Mast cells&quot;</span><span class="p">,</span>
<span class="p">}</span>


<span class="n">sdata_crop</span><span class="o">.</span><span class="n">tables</span><span class="p">[</span><span class="s2">&quot;table&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;celltype&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span>
    <span class="n">sdata_crop</span><span class="o">.</span><span class="n">tables</span><span class="p">[</span><span class="s2">&quot;table&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;leiden&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="n">celltype_mapping</span><span class="p">)</span>
<span class="p">)</span>
<span class="n">sdata_crop</span><span class="o">.</span><span class="n">tables</span><span class="p">[</span><span class="s2">&quot;table&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;celltype&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">fillna</span><span class="p">(</span><span class="s2">&quot;Unknown&quot;</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

<span class="n">sdata_crop</span><span class="o">.</span><span class="n">tables</span><span class="p">[</span><span class="s2">&quot;table&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;celltype&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">Categorical</span><span class="p">(</span>
    <span class="n">sdata_crop</span><span class="o">.</span><span class="n">tables</span><span class="p">[</span><span class="s2">&quot;table&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;celltype&quot;</span><span class="p">]</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/var/folders/qg/qgc908995g3fc8qtss2fsbhhxyxxj4/T/ipykernel_63128/1239269867.py:17: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing &#39;df[col].method(value, inplace=True)&#39;, try using &#39;df.method({col: value}, inplace=True)&#39; or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  sdata_crop.tables[&quot;table&quot;].obs[&quot;celltype&quot;].fillna(&quot;Unknown&quot;, inplace=True)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">fig</span><span class="p">,</span> <span class="n">axs</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">15</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>

<span class="n">sdata_crop</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">render_images</span><span class="p">(</span><span class="s2">&quot;he_image&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">show</span><span class="p">(</span><span class="n">ax</span><span class="o">=</span><span class="n">axs</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">title</span><span class="o">=</span><span class="s2">&quot;H&amp;E&quot;</span><span class="p">)</span>

<span class="n">sdata_crop</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">render_shapes</span><span class="p">(</span>
    <span class="s2">&quot;cell_circles&quot;</span><span class="p">,</span>
    <span class="n">color</span><span class="o">=</span><span class="s2">&quot;celltype&quot;</span><span class="p">,</span>
    <span class="n">method</span><span class="o">=</span><span class="s2">&quot;matplotlib&quot;</span><span class="p">,</span>
<span class="p">)</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">show</span><span class="p">(</span><span class="n">ax</span><span class="o">=</span><span class="n">axs</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">title</span><span class="o">=</span><span class="s2">&quot;Leiden&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Clipping input data to the valid range for imshow with RGB data ([0..1] for floats or [0..255] for integers). Got range [-0.053719006..1.0].
/Users/tim.treis/anaconda3/envs/spatialdata-workshop/lib/python3.11/site-packages/spatialdata/_core/_elements.py:105: UserWarning: Key `cell_circles` already exists. Overwriting it in-memory.
  self._check_key(key, self.keys(), self._shared_keys)
/Users/tim.treis/anaconda3/envs/spatialdata-workshop/lib/python3.11/site-packages/spatialdata/_core/_elements.py:125: UserWarning: Key `table` already exists. Overwriting it in-memory.
  self._check_key(key, self.keys(), self._shared_keys)
/Users/tim.treis/anaconda3/envs/spatialdata-workshop/lib/python3.11/site-packages/spatialdata_plot/pl/utils.py:771: FutureWarning: The default value of &#39;ignore&#39; for the `na_action` parameter in pandas.Categorical.map is deprecated and will be changed to &#39;None&#39; in a future version. Please set na_action to the desired value to avoid seeing this warning
  color_vector = color_source_vector.map(color_mapping)
</pre></div>
</div>
<img alt="../../../_images/b144e774d230a65700f9793eb20dadb198f9a276a28baf30b1ced9afc6492049.png" src="../../../_images/b144e774d230a65700f9793eb20dadb198f9a276a28baf30b1ced9afc6492049.png" />
</div>
</div>
<p>We can also try to overlay these plots.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">fig</span><span class="p">,</span> <span class="n">axs</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">15</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>

<span class="n">sdata_crop</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">render_images</span><span class="p">(</span>
    <span class="s2">&quot;he_image&quot;</span><span class="p">,</span>
    <span class="c1"># cmap=&quot;Greys&quot;,  # since the cmap argument currently has a bug with multiscale images, we&#39;ll use a simple grey colorscale</span>
    <span class="n">channel</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>
    <span class="n">palette</span><span class="o">=</span><span class="s2">&quot;lightgrey&quot;</span><span class="p">,</span>
<span class="p">)</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">render_shapes</span><span class="p">(</span>
    <span class="s2">&quot;cell_circles&quot;</span><span class="p">,</span>
    <span class="n">color</span><span class="o">=</span><span class="s2">&quot;celltype&quot;</span><span class="p">,</span>
    <span class="n">fill_alpha</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span>
    <span class="n">method</span><span class="o">=</span><span class="s2">&quot;matplotlib&quot;</span><span class="p">,</span>
<span class="p">)</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">show</span><span class="p">(</span>
    <span class="n">ax</span><span class="o">=</span><span class="n">axs</span><span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="s2">&quot;Celltypes&quot;</span><span class="p">,</span> <span class="n">colorbar</span><span class="o">=</span><span class="kc">False</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/Users/tim.treis/anaconda3/envs/spatialdata-workshop/lib/python3.11/site-packages/spatialdata/_core/_elements.py:105: UserWarning: Key `cell_circles` already exists. Overwriting it in-memory.
  self._check_key(key, self.keys(), self._shared_keys)
/Users/tim.treis/anaconda3/envs/spatialdata-workshop/lib/python3.11/site-packages/spatialdata/_core/_elements.py:125: UserWarning: Key `table` already exists. Overwriting it in-memory.
  self._check_key(key, self.keys(), self._shared_keys)
/Users/tim.treis/anaconda3/envs/spatialdata-workshop/lib/python3.11/site-packages/spatialdata_plot/pl/utils.py:771: FutureWarning: The default value of &#39;ignore&#39; for the `na_action` parameter in pandas.Categorical.map is deprecated and will be changed to &#39;None&#39; in a future version. Please set na_action to the desired value to avoid seeing this warning
  color_vector = color_source_vector.map(color_mapping)
</pre></div>
</div>
<img alt="../../../_images/073d3bbc3f518f3225bf259970d74e532e56857514bbbd75eac5db084431ba21.png" src="../../../_images/073d3bbc3f518f3225bf259970d74e532e56857514bbbd75eac5db084431ba21.png" />
</div>
</div>
</section>
</section>
<section id="studying-the-cell-type-fingerprint-for-regions-of-interest">
<h2>Studying the cell-type fingerprint for regions of interest<a class="headerlink" href="#studying-the-cell-type-fingerprint-for-regions-of-interest" title="Link to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">napari_spatialdata</span><span class="w"> </span><span class="kn">import</span> <span class="n">Interactive</span>

<span class="c1"># uncomment to annotate the tiles/regions of interest manually</span>
<span class="c1"># Interactive(sdata_crop)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">shapely</span><span class="w"> </span><span class="kn">import</span> <span class="n">Polygon</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">geopandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">gpd</span>

<span class="n">coords0</span> <span class="o">=</span> <span class="p">[(</span><span class="mi">31892</span><span class="p">,</span> <span class="mi">6628</span><span class="p">),</span> <span class="p">(</span><span class="mi">33814</span><span class="p">,</span> <span class="mi">6628</span><span class="p">),</span> <span class="p">(</span><span class="mi">33814</span><span class="p">,</span> <span class="mi">8484</span><span class="p">),</span> <span class="p">(</span><span class="mi">31892</span><span class="p">,</span> <span class="mi">8484</span><span class="p">),</span> <span class="p">(</span><span class="mi">31892</span><span class="p">,</span> <span class="mi">6628</span><span class="p">)]</span>
<span class="n">coords1</span> <span class="o">=</span> <span class="p">[(</span><span class="mi">30336</span><span class="p">,</span> <span class="mi">9126</span><span class="p">),</span> <span class="p">(</span><span class="mi">32258</span><span class="p">,</span> <span class="mi">9126</span><span class="p">),</span> <span class="p">(</span><span class="mi">32258</span><span class="p">,</span> <span class="mi">10982</span><span class="p">),</span> <span class="p">(</span><span class="mi">30336</span><span class="p">,</span> <span class="mi">10982</span><span class="p">),</span> <span class="p">(</span><span class="mi">30336</span><span class="p">,</span> <span class="mi">9126</span><span class="p">)]</span>
<span class="n">coords2</span> <span class="o">=</span> <span class="p">[(</span><span class="mi">26948</span><span class="p">,</span> <span class="mi">6706</span><span class="p">),</span> <span class="p">(</span><span class="mi">28870</span><span class="p">,</span> <span class="mi">6706</span><span class="p">),</span> <span class="p">(</span><span class="mi">28870</span><span class="p">,</span> <span class="mi">8562</span><span class="p">),</span> <span class="p">(</span><span class="mi">26948</span><span class="p">,</span> <span class="mi">8562</span><span class="p">),</span> <span class="p">(</span><span class="mi">26948</span><span class="p">,</span> <span class="mi">6706</span><span class="p">)]</span>

<span class="n">poly0</span> <span class="o">=</span> <span class="n">Polygon</span><span class="p">(</span><span class="n">coords0</span><span class="p">)</span>
<span class="n">poly1</span> <span class="o">=</span> <span class="n">Polygon</span><span class="p">(</span><span class="n">coords1</span><span class="p">)</span>
<span class="n">poly2</span> <span class="o">=</span> <span class="n">Polygon</span><span class="p">(</span><span class="n">coords2</span><span class="p">)</span>
<span class="n">gdf</span> <span class="o">=</span> <span class="n">sd</span><span class="o">.</span><span class="n">models</span><span class="o">.</span><span class="n">ShapesModel</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span>
    <span class="n">gpd</span><span class="o">.</span><span class="n">GeoDataFrame</span><span class="p">({</span><span class="s2">&quot;geometry&quot;</span><span class="p">:</span> <span class="p">[</span><span class="n">poly0</span><span class="p">,</span> <span class="n">poly1</span><span class="p">,</span> <span class="n">poly2</span><span class="p">]}),</span>
    <span class="n">transformations</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;global&quot;</span><span class="p">:</span> <span class="n">sd</span><span class="o">.</span><span class="n">transformations</span><span class="o">.</span><span class="n">Identity</span><span class="p">()},</span>
<span class="p">)</span>

<span class="n">sdata_crop</span><span class="p">[</span><span class="s2">&quot;tiles&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">gdf</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sdata_crop</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>SpatialData object
├── Images
│     ├── &#39;he_image&#39;: DataTree[cyx] (3, 973, 973), (3, 487, 487), (3, 243, 243), (3, 122, 122), (3, 60, 61)
│     └── &#39;morphology_focus&#39;: DataTree[cyx] (1, 10000, 10000), (1, 5000, 5000), (1, 2500, 2499), (1, 1250, 1250), (1, 625, 625)
├── Labels
│     ├── &#39;cell_labels&#39;: DataTree[yx] (10000, 10000), (5000, 5000), (2500, 2499), (1250, 1250), (625, 625)
│     └── &#39;nucleus_labels&#39;: DataTree[yx] (10000, 10000), (5000, 5000), (2500, 2499), (1250, 1250), (625, 625)
├── Points
│     └── &#39;transcripts&#39;: DataFrame with shape: (&lt;Delayed&gt;, 11) (3D points)
├── Shapes
│     ├── &#39;cell_boundaries&#39;: GeoDataFrame shape: (33886, 1) (2D shapes)
│     ├── &#39;cell_circles&#39;: GeoDataFrame shape: (33727, 2) (2D shapes)
│     ├── &#39;nucleus_boundaries&#39;: GeoDataFrame shape: (32342, 1) (2D shapes)
│     └── &#39;tiles&#39;: GeoDataFrame shape: (3, 1) (2D shapes)
└── Tables
      └── &#39;table&#39;: AnnData (33727, 377)
with coordinate systems:
    ▸ &#39;global&#39;, with elements:
        he_image (Images), morphology_focus (Images), cell_labels (Labels), nucleus_labels (Labels), transcripts (Points), cell_boundaries (Shapes), cell_circles (Shapes), nucleus_boundaries (Shapes), tiles (Shapes)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cell_types_fractions</span> <span class="o">=</span> <span class="n">sd</span><span class="o">.</span><span class="n">aggregate</span><span class="p">(</span>
    <span class="n">values_sdata</span><span class="o">=</span><span class="n">sdata_crop</span><span class="p">,</span>
    <span class="n">values</span><span class="o">=</span><span class="s2">&quot;cell_circles&quot;</span><span class="p">,</span>
    <span class="n">by_sdata</span><span class="o">=</span><span class="n">sdata_crop</span><span class="p">,</span>
    <span class="n">by</span><span class="o">=</span><span class="s2">&quot;tiles&quot;</span><span class="p">,</span>
    <span class="n">value_key</span><span class="o">=</span><span class="s2">&quot;celltype&quot;</span><span class="p">,</span>
    <span class="n">target_coordinate_system</span><span class="o">=</span><span class="s2">&quot;global&quot;</span><span class="p">,</span>
    <span class="n">fractions</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">cell_types_fractions</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/Users/tim.treis/anaconda3/envs/spatialdata-workshop/lib/python3.11/site-packages/spatialdata/_core/operations/aggregate.py:453: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  aggregated = joined.groupby([INDEX, vk])[ONES_COLUMN].agg(agg_func).reset_index()
/Users/tim.treis/anaconda3/envs/spatialdata-workshop/lib/python3.11/site-packages/anndata/_core/anndata.py:381: FutureWarning: The dtype argument is deprecated and will be removed in late 2024.
  warnings.warn(
/Users/tim.treis/anaconda3/envs/spatialdata-workshop/lib/python3.11/site-packages/anndata/_core/aligned_df.py:68: ImplicitModificationWarning: Transforming to str index.
  warnings.warn(&quot;Transforming to str index.&quot;, ImplicitModificationWarning)
/Users/tim.treis/anaconda3/envs/spatialdata-workshop/lib/python3.11/site-packages/spatialdata/models/models.py:1053: UserWarning: Converting `region_key: region` to categorical dtype.
  return convert_region_column_to_categorical(adata)
</pre></div>
</div>
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>SpatialData object
├── Shapes
│     └── &#39;tiles&#39;: GeoDataFrame shape: (3, 1) (2D shapes)
└── Tables
      └── &#39;table&#39;: AnnData (3, 8)
with coordinate systems:
    ▸ &#39;global&#39;, with elements:
        tiles (Shapes)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sdata_crop</span><span class="p">[</span><span class="s2">&quot;tiles&quot;</span><span class="p">][</span><span class="s2">&quot;name&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="sa">f</span><span class="s2">&quot;tile_</span><span class="si">{</span><span class="n">i</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">sdata_crop</span><span class="p">[</span><span class="s2">&quot;tiles&quot;</span><span class="p">]))]</span>
<span class="n">sdata_crop</span><span class="p">[</span><span class="s2">&quot;tiles&quot;</span><span class="p">][</span><span class="s2">&quot;name&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">sdata_crop</span><span class="p">[</span><span class="s2">&quot;tiles&quot;</span><span class="p">][</span><span class="s2">&quot;name&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s2">&quot;category&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">np</span>

<span class="n">fig</span><span class="p">,</span> <span class="n">axes</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">7</span><span class="p">,</span> <span class="mi">10</span><span class="p">))</span>

<span class="c1"># plot the tiles</span>
<span class="p">(</span>
    <span class="n">sdata_crop</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">render_images</span><span class="p">(</span><span class="s2">&quot;he_image&quot;</span><span class="p">)</span>
    <span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">render_shapes</span><span class="p">(</span><span class="s2">&quot;tiles&quot;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;name&quot;</span><span class="p">,</span> <span class="n">fill_alpha</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span> <span class="n">outline_alpha</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
    <span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">show</span><span class="p">(</span><span class="n">ax</span><span class="o">=</span><span class="n">axes</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
<span class="p">)</span>

<span class="c1"># plot the cell type fractions</span>
<span class="n">roi_names</span> <span class="o">=</span> <span class="n">sdata_crop</span><span class="p">[</span><span class="s2">&quot;tiles&quot;</span><span class="p">][</span><span class="s2">&quot;name&quot;</span><span class="p">]</span>
<span class="n">celltype_names</span> <span class="o">=</span> <span class="n">cell_types_fractions</span><span class="o">.</span><span class="n">table</span><span class="o">.</span><span class="n">var_names</span>

<span class="n">data</span> <span class="o">=</span> <span class="n">cell_types_fractions</span><span class="o">.</span><span class="n">table</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">todense</span><span class="p">()</span><span class="o">.</span><span class="n">A</span>
<span class="n">data_normalized</span> <span class="o">=</span> <span class="n">data</span> <span class="o">/</span> <span class="n">data</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)[:,</span> <span class="n">np</span><span class="o">.</span><span class="n">newaxis</span><span class="p">]</span>
<span class="n">colors</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">cm</span><span class="o">.</span><span class="n">tab10</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">linspace</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">10</span><span class="p">))</span>

<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">data</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]):</span>
    <span class="n">left</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="k">for</span> <span class="n">j</span><span class="p">,</span> <span class="n">name</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">celltype_names</span><span class="p">):</span>
        <span class="n">axes</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">barh</span><span class="p">(</span>
            <span class="n">i</span><span class="p">,</span>
            <span class="n">data_normalized</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="n">j</span><span class="p">],</span>
            <span class="n">left</span><span class="o">=</span><span class="n">left</span><span class="p">,</span>
            <span class="n">color</span><span class="o">=</span><span class="n">colors</span><span class="p">[</span><span class="n">j</span><span class="p">],</span>
            <span class="n">label</span><span class="o">=</span><span class="n">name</span> <span class="k">if</span> <span class="n">i</span> <span class="o">==</span> <span class="mi">0</span> <span class="k">else</span> <span class="s2">&quot;&quot;</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="n">left</span> <span class="o">+=</span> <span class="n">data_normalized</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="n">j</span><span class="p">]</span>

<span class="n">axes</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_yticks</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="n">data</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]))</span>
<span class="n">axes</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_yticklabels</span><span class="p">([</span><span class="n">roi_names</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">data</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">])])</span>

<span class="n">axes</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_xticks</span><span class="p">([])</span>

<span class="n">handles</span><span class="p">,</span> <span class="n">labels</span> <span class="o">=</span> <span class="n">axes</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">get_legend_handles_labels</span><span class="p">()</span>
<span class="n">by_label</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">labels</span><span class="p">,</span> <span class="n">handles</span><span class="p">))</span>  <span class="c1"># This removes duplicate labels/handles</span>
<span class="n">axes</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span>
    <span class="n">by_label</span><span class="o">.</span><span class="n">values</span><span class="p">(),</span> <span class="n">by_label</span><span class="o">.</span><span class="n">keys</span><span class="p">(),</span> <span class="n">loc</span><span class="o">=</span><span class="s2">&quot;upper left&quot;</span><span class="p">,</span> <span class="n">bbox_to_anchor</span><span class="o">=</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
<span class="p">)</span>

<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;Xenium cell-type fractions for ROIs&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Clipping input data to the valid range for imshow with RGB data ([0..1] for floats or [0..255] for integers). Got range [-0.053719006..1.0].
/Users/tim.treis/anaconda3/envs/spatialdata-workshop/lib/python3.11/site-packages/spatialdata_plot/pl/utils.py:771: FutureWarning: The default value of &#39;ignore&#39; for the `na_action` parameter in pandas.Categorical.map is deprecated and will be changed to &#39;None&#39; in a future version. Please set na_action to the desired value to avoid seeing this warning
  color_vector = color_source_vector.map(color_mapping)
/var/folders/qg/qgc908995g3fc8qtss2fsbhhxyxxj4/T/ipykernel_63128/2377721173.py:14: DeprecationWarning: Table accessor will be deprecated with SpatialData version 0.1, use sdata.tables instead.
  celltype_names = cell_types_fractions.table.var_names
/var/folders/qg/qgc908995g3fc8qtss2fsbhhxyxxj4/T/ipykernel_63128/2377721173.py:16: DeprecationWarning: Table accessor will be deprecated with SpatialData version 0.1, use sdata.tables instead.
  data = cell_types_fractions.table.X.todense().A
</pre></div>
</div>
<img alt="../../../_images/4420e489a399de61ac98193ec66815ebfb43d7ad38f2e8bceae020aa40a56003.png" src="../../../_images/4420e489a399de61ac98193ec66815ebfb43d7ad38f2e8bceae020aa40a56003.png" />
</div>
</div>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            name: "python3",
            path: "./day_2/spatialdata/beginner"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

                </article>
              

              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#subset-the-xenium-data-for-the-scope-of-this-workshop">Subset the Xenium data for the scope of this workshop</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#investigate-the-proposed-leiden-clusters">Investigate the proposed Leiden clusters</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#rename-clusters">Rename clusters</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#studying-the-cell-type-fingerprint-for-regions-of-interest">Studying the cell-type fingerprint for regions of interest</a></li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By The Jupyter Book community
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
      © Copyright 2023.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../../../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b"></script>
<script src="../../../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>